### DISCUSSION TITLE
Reference DataType

### DISCUSSION THREAD

**Discussion by cnash** [jamesmudd/jhdf#311] [Q&A] [answered]
I'm trying to pull some data from a Matlab MAT-file v7.3. Some of it is coming back with a data type of "Reference". I'm having a heck of a time figuring out how to de-reference those. Is there anyway you could point me to some sample code?

--Thanks!

**Comment by jamesmudd** [comment]
Hi. Unfortunately I don't think I can point to a "nice" example at the moment. `Reference` contain a `long` which is a address in the file to another object. Currently there is not a "nice" API to dereference them, but i'm happy to look at that as an improvement. Do you have an example file you would be willing to share? Would be interested to see how its used.

A possible workaround would be to walk the file e.g. see https://github.com/jamesmudd/jhdf/blob/master/jhdf/src/main/java/io/jhdf/examples/PrintTree.java#L45 you could change `node.getName()` to `node.getAddress()` and find the node with the address the reference points to. At that point you have the `Node` so you can call `getData()` assuming its a dataset. (This depends a bit on exactly what the references are in your file so it might not work that easily for you)

**Reply by cnash** [reply]
I was only given one MAT-file to work from, and I'm told it contains proprietary info that I can't share.  I'll try to find one that I can share.

**Comment by jamesmudd** [comment, marked as answer]
Some untested code that implements the workaround
```
public Object getDataFromReference(HdfFile hdfFile, long address) {
	return ((Dataset) recurseGroup(hdfFile, address)).getData();
}

private Node recurseGroup(Group group, long address) {
	for (Node node : group) {
		if(node.getAddress() == address)  {
			return node;
		} else if (node instanceof Group) {
			recurseGroup((Group) node, address);
		}
	}
	throw new RuntimeException("Didn't find reference");
}
```

**Reply by cnash** [reply]
Thanks! I'll try this out. 
I was able to get the "#refs# Group from the file, and then search through them for nodes having the appropriate Address. I'll share the solution once I've cleaned it up.

**Reply by cnash** [reply]
I modified your code slightly. I think it would have caused unhandled exceptions from the recursion.  Here's what I have now and it seems to work.

```
private Node dereference(Group group, long address) {
	for (Node node : group) {
		if(node.getAddress() == address)  {
			return node;
		} else if (node instanceof Group) {
			Node foundNode = dereference((Group) node, address);
			if (foundNode != null) return foundNode;
		}
	}
	return null;
}
```

**Comment by jamesmudd** [comment]
As of [v0.6.8](https://github.com/jamesmudd/jhdf/releases/tag/v0.6.8) there is now a new method `HdfFile.getNodeByAddress(long address)` which offers a nicer solution to this that searching through the file with much better performance.
